pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/AuthenticationProviderTokenOIDC.java

```diff
@@ -0,0 +1,361 @@
+            // Authenticate Pulsar binary connection
+            return validateToken(authData.getCommandData());
+        } else if (authData.hasDataFromHttp()) {
+            // Authentication HTTP request. The format here should be compliant to RFC-6750
+            // (https://tools.ietf.org/html/rfc6750#section-2.1). Eg: Authorization: Bearer xxxxxxxxxxxxx
+            String httpHeaderValue = authData.getHttpHeader(HTTP_HEADER_NAME);
+            if (httpHeaderValue == null || !httpHeaderValue.startsWith(HTTP_HEADER_VALUE_PREFIX)) {
+                throw new AuthenticationException("Invalid HTTP Authorization header");
+            }
+
+            // Remove prefix
+            String token = httpHeaderValue.substring(HTTP_HEADER_VALUE_PREFIX.length());
+            return validateToken(token);
+        } else {
+            throw new AuthenticationException("No token credentials passed");
+        }
+    }
+
+    private static String validateToken(final String token) throws AuthenticationException {
+        if (StringUtils.isNotBlank(token)) {
+            return token;
+        } else {
+            throw new AuthenticationException("Blank token found");
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private DecodedJWT authenticateToken(final String token) throws AuthenticationException {
+        DecodedJWT jwt = JWT.decode(token);
+
+        Jwk jwk = null;
+        Algorithm algorithm = null;
+        try {
+            jwk = provider.get(jwt.getKeyId());
+            if(!(jwk.getPublicKey() instanceof RSAPublicKey)){
+                throw new JwtException("key needs to be a RSA Publickey");
+            }
+
+            algorithm = Algorithm.RSA256((RSAPublicKey) jwk.getPublicKey(), null);
+            algorithm.verify(jwt); // if the token signature is invalid, the method will throw SignatureVerificationException
+
+            if (audienceClaim != null) {
+                Object object = jwt.getClaim(audienceClaim);
+
+                if (object == null) {
+                    throw new JwtException("Found null Audience in token, for claimed field: " + audienceClaim);
+                }
+
+                if (object instanceof List) {
+                    List<String> audiences = (List<String>) object;
+                    // audience not contains this broker, throw exception.
+                    if (audiences.stream().noneMatch(audienceInToken -> audienceInToken.equals(audience))) {
+                        throw new AuthenticationException("Audiences in token: [" + String.join(", ", audiences)
+                                + "] not contains this broker: " + audience);
+                    }
+                } else if (object instanceof String) {
+                    if (!object.equals(audience)) {
+                        throw new AuthenticationException("Audiences in token: [" + object
+                                + "] not contains this broker: " + audience);
```

Comment:
Regarding the error message, `not contains this broker`, don't assume that the audience represents the broker because that's an oauth2 thing.  It might simply represent the allowed users.

To be honest, if you want to support audience validation in an OIDC scenario, the broker configuration should accept a list of allowed audiences.